openage/convert/processor/export/modpack_manifest.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2020-2020 the openage authors. See copying.md for legal info.
+
+"""
+Create a manifest file for a modpack
+"""
+import hashlib
+import os
+import toml
+
+from ....util.fslike.path import Path
+from ...entity_object.export.data_definition import DataDefinition
+
+
+class ManifestFile(DataDefinition):
+    """
+    Used for creating a manifest file for a modpack.
+    """
+
+    HASHING_FUNCTION = 'sha3_256'
+    BUF_SIZE = 32000
+
+    def __init__(self, targetdir, filename):
+        super().__init__(targetdir, filename)
+
+    def dump(self, exportdir):
+        """
+        Returns the manifest file content in TOML format.
+
+        :param exportdir: directory which contains exported modpack
+        :type exportdir: ....util.fslike.path.Path
+        """
+
+        output_dict = {}
+
+        info_table = {"info": {}}
+        info_table["info"].update({"hash": self.HASHING_FUNCTION})
+
+        output_dict.update(info_table)
+
+        hash_values_table = {'hash-values': {}}
+        for file in self._bfs_directory(exportdir):
+            hash_val = self._get_file_hash(file)
+            relative_path = os.path.relpath(str(file), str(exportdir))
+            hash_values_table['hash-values'].update({hash_val: relative_path})
```

Comment:
We would like to separate the calculation of the hashes from saving the hashes. This means that the `ManifestFile` does not do the BFS search and hash calculation, but is instead initialized with an empty list that stores the hash values and a method that adds a tuple `(hash, filepath)` to the list. The function doing the BFS search then just appends a hash value using the method.

Instances of `DataDefinition` should generally only get the data they dump and not be involved in the generation of said data. The reason for this is that, as a result, data can be generated and added from multiple sources. `ManifestFile` should be a storage for data that is exported. It does not need to know where the data is from or how it was generated.

Reply:
Ok yeah understood. So `_bfs_directory` and `_get_file_hash` should be in a separate processor and `ManifestFile` should use a method which actually uses both of them to generate a `(hash, filepath)` tuple. I get that right?

But then after separating the functions should the `ManifestFile` still be in the processor subdirectory? As I made it an instance of `DataDefinition` should it be now placed at `entity_object` subdirectory?

Reply:
> `ManifestFile` should use a method which actually uses both of them to generate a (hash, filepath) tuple. I get that right?

`ManifestFile` should have a method that takes a `(hash, filepath)` and puts them into a list. The processor should generate them using `_bfs_directory` and `_get_file_hash` and pass them to `ManifestFile` by using the method.

> But then after separating the functions should the ManifestFile still be in the processor subdirectory?

`ManifestFile` goes into `openage/convert/entity_object/export/formats` where all the other format classes are. The `_bfs_directory` and `_get_file_hash` functions should stay in this file.